api/renter_test.go

```diff
@@ -1899,3 +1899,212 @@ func TestContractorHostRemoval(t *testing.T) {
 		t.Log("downloaded file and uploaded file do not match")
 	}
 }
+
+// TestExhaustedContracts verifies that the contractor renews contracts which
+// run out of funds before the period elapses.
+func TestExhaustedContracts(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+	t.Parallel()
+	st, err := createServerTester(t.Name())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer st.server.panicClose()
+
+	// set a very high price for the host so we use up the entire funds of the
+	// contract
+	err = st.acceptContracts()
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = st.setHostStorage()
+	if err != nil {
+		t.Fatal(err)
+	}
+	settings := st.host.InternalSettings()
+	settings.MinUploadBandwidthPrice = types.SiacoinPrecision.Div64(10)
+	err = st.host.SetInternalSettings(settings)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = st.announceHost()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Set an allowance for the renter, allowing a contract to be formed.
+	allowanceValues := url.Values{}
+	testPeriod := "1000"
+	allowanceValues.Set("funds", types.SiacoinPrecision.Mul64(1000).String())
+	allowanceValues.Set("period", testPeriod)
+	err = st.stdPostAPI("/renter", allowanceValues)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// wait until we have a contract
+	err = retry(500, time.Millisecond*50, func() error {
+		if len(st.renter.Contracts()) >= 1 {
+			return nil
+		}
+		return errors.New("no renter contracts")
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	initialContract := st.renter.Contracts()[0]
+
+	// upload a file. the high upload cost will cause the underlying contract to
+	// require premature renewal. If premature renewal never happens, the upload
+	// will never complete.
+	tmpfile, err := ioutil.TempFile("", t.Name())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.Remove(tmpfile.Name())
+	_, err = io.CopyN(tmpfile, fastrand.Reader, int64(modules.SectorSize))
```

Comment:
I'm not sure if I'm missing something, but I think you would want several sectors, where the contract has enough funds to upload a few sectors, but not all of them. Right now it looks like it just uploads one sector, and I'm not sure how an upload could get stuck halfway through if there's only one host and one sector.

Reply:
This test sets the upload bandwidth cost very high, high enough so that the contract gets exhausted part-way through uploading just one sector.